### PR TITLE
fix UB in file_checksum method

### PR DESCRIPTION
**Pull request by ailin-nemui**
remove UB on high chars

### CODE REVIEW

**Comment by LemonBoy** [comment]
The commit message is atrocious but #836 takes the cake.

**Comment by dequis** [comment]
c0f66c95ff00bdc22155a8ac53d887a9c1bbc90f never forget

**Comment by josephbisch** [comment]
After applying this patch:

```
settings.c:678:25: runtime error: left shift of 138 by 24 places cannot be represented in type 'int'
```

**Comment by ailin-nemui** [comment]
how about something real instead

**Comment by LemonBoy** [comment]
~~I can't really wait to add more of OpenSSL in the mix~~

We can use [glib](https://developer.gnome.org/glib/stable/glib-Data-Checksums.html) instead since we bumped the minimum version to 2.28 some time ago.

**Comment by dequis** [comment]
Yeah I thought the same thing but really I have no opinion on glib checksums vs openssl. The libcrypto parts of openssl are the safest and hashing is a relatively trivial problem.

`g_compute_checksum_for_data` would be shorter but heap allocated so to keep the same allocation style you would end up going through the `g_checksum_new` route which takes the same amount of code. Not that heap allocation hurts.

Also obligatory mention that sha1 has collisions even though the older checksum was worse lol.

**Comment by ailin-nemui** [comment]
@LemonBoy want to rewrite it with glib? I don't care too much which of the libs we use

**Comment by ailin-nemui** [comment]
@irssi/developers

**Comment by dequis** [comment]
Hm, why did you choose not to use `g_compute_checksum_for_data`?

**Comment by ailin-nemui** [comment]
for that we need to read the whole file into memory first

**Comment by dequis** [comment]
Fair

File: src/core/settings.c [hunk 1]
```diff
@@ -661,24 +662,44 @@ void sig_term(int n)
...
 	while ((ret = read(f, buf, sizeof(buf))) > 0) {
-		while (ret-- > 0)
-			checksum += buf[ret] << ((n++ & 3)*8);
+		g_checksum_update(c, buf, ret);
 	}
 	close(f);
-	return checksum;
+
+	checksum = g_new0(guint8, g_checksum_type_get_length(checksum_type));
+	g_checksum_get_digest(c, checksum, &checksum_length);
```
Comment (dequis) [inline src/core/settings.c]: This crashes with valgrind (and gets good enough garbage in the stack to work decently the rest of the time)

```
==28250== Conditional jump or move depends on uninitialised value(s)
==28250==    at 0x56A2855: g_checksum_get_digest (in /usr/lib/libglib-2.0.so.0.5400.0)
==28250==    by 0x1B1AAE: file_checksum (settings.c:689)
==28250==    by 0x1B1BC1: irssi_config_save_state (settings.c:717)
==28250==    by 0x1B1C57: parse_configfile (settings.c:774)
==28250==    by 0x1B3310: init_configfile (settings.c:794)
==28250==    by 0x1B3310: settings_init (settings.c:901)
==28250==    by 0x19FE44: core_init (core.c:237)
==28250==    by 0x12BE0F: textui_init (irssi.c:147)
==28250==    by 0x12BE0F: main (irssi.c:310)

(irssi:28250): GLib-CRITICAL **: g_checksum_get_digest: assertion '*digest_len >= len' failed
zsh: trace trap (core dumped)  valgrind --log-file=leg ./src/fe-text/irssi
```

checksum_length is inout, so initialize it to `g_checksum_type_get_length(checksum_type)`

Comment (ailin-nemui) [reply src/core/settings.c]: great

**Comment by ailin-nemui** [comment]
I don't like inout parameters...

**Pull request closed without merging**